pytket/extensions/cutensornet/tnstate/ttn_gate.py

```diff
@@ -258,27 +258,168 @@ def _apply_2q_gate(self, q0: Qubit, q1: Qubit, gate: Op) -> TTNxGate:
+                normalization="L2",  # Sum of squares singular values must equal 1
+            )
+
+            # Apply the SVD decomposition using the configuration defined above
+            U, S, V, svd_info = tensor.decompose(
+                "cp->cs,sp",
+                bond_tensor,
+                method=svd_method,
+                options=options,
+                return_info=True,
+            )
+            assert S is None  # Due to "partition" option in SVDMethod
+
+            # discarded_weight is calculated within cuTensorNet as:
+            #                             sum([s**2 for s in S'])
+            #     discarded_weight = 1 - -------------------------
+            #                             sum([s**2 for s in S])
+            # where S is the list of original singular values and S' is the set of
+            # singular values that remain after truncation (before normalisation).
+            # It can be shown that the fidelity |<psi|phi>|^2 (for |phi> and |psi>
+            # unit vectors before and after truncation) is equal to 1 - disc_weight.
+            #
+            # We multiply the fidelity of the current step to the overall fidelity
+            # to keep track of a lower bound for the fidelity.
+            trunc_fidelity = 1.0 - svd_info.discarded_weight
+            self.fidelity *= trunc_fidelity
+            dimension_after = V.shape[0]
+
+            # Report to logger
+            self._logger.debug(f"Truncation done. Truncation fidelity={trunc_fidelity}")
+            self._logger.debug(
+                f"Reduced bond dimension from {dimension_before} to {dimension_after}."
+            )
+
+            # We require that self._cfg.truncation_fidelity > prod(each_trunc_fidelity)
+            # since the RHS lower bounds the actual update factor to the fidelity. We
+            # ensure this inequality is satisfied by dividing the fidelity_threshold
+            # by the fidelity of this truncation. This makes the threshold tighter,
+            # since trunc_fidelity is less than 1, so division causes a higher value
+            # of fidelity_threshold.
+            fidelity_threshold /= trunc_fidelity
+
+            # Contract U and V into the TTN. This reintroduces the data of bond_tensor
+            # back into the TTN, as required by ``canonicalise(.., unsafe=True)``.
+            self._contract_decomp_bond_tensor_into_ttn(U, V, bond_address)
+
+            # Since S was contracted with U, and the U contracted with the child node,
+            # the child node loses its canonical form.
+            self.nodes[bond_address].canonical_form = None
+
+            # Stop truncating if there was no reduction of dimension in this iteration
+            if dimension_before == dimension_after:
+                break
+
+    def _chi_sequential_bond_truncation(
+        self,
+        truncation_bonds: list[RootPath],
+    ) -> None:

```

Comment:
This is the existing method (from branch `feature/ttn`) that truncates to a given value of `chi`. I've moved code around so that it's more readable and modular, but the logic itself has not changed.